content/ADR-146-unity-2021-material-upgrade.md

```diff
@@ -0,0 +1,77 @@
+
+## Upgrading materials from 2020 to 2021
+Unfortunately Unity is using the same class to serialize materials from Unity 2020 to Unity 2021, that means that we have no access to anything the new material doesnt provide. That makes finding stuff like the rendering target (Opaque, Alpha Test, Transparency) challenging because the new URP changed the way keywords are serialized. 
+
+Luckily the shader properties are still there, so we can gather them to try to find the correct setup for the material. Additionally we optimize the materials for the SRP Batcher to enhance performance, this optimizations give us some insight of the render target of the original material.
+
+### Getting Render target through the SRP Batcher optimizations
+SRP Batcher itself is a whole new world that fall outside the scope of this ADR. The minimum we have to know about it is that optimizes draw calls by batching objects being rendered with the same shader variant. A common variant example is having `_ALPHATEST_ON` enabled or not to clip fragments that are invisible based on an alpha value. One of the optimizations we do is assume every opaque material is alpha tested and set the property `_CutOff` (the alpha value for clipping) to `0` in the cases we dont want to clip.
+
+The relevant piece that will solve our problem is tied to the Render Queue. If our render queue is displaying something like:
+- MaterialA_variantA 2050
+- MaterialA_variantB 2051
+- MaterialA_variantA 2052
+
+SRP Batcher wont be able to batch `MaterialA_variantA` because it has to render `MaterialA_variantB` in the middle. We try to put materials with the same render target (opaque, alphatested, transparent) and the same properties (same variant) in the same render queue and we will use this to know more about the original material.
+
+
+The full details of the algorithm can be found [here](https://github.com/decentraland/unity-renderer/blob/dev/unity-renderer/Assets/Scripts/MainScripts/DCL/Helpers/MaterialHelpers/SRPBatchingHelper/SRPBatchingHelper.cs) but the conclusion is that we are gonna have this range of render queues depending on the render target:
+
+```
+Opaque 2000 + crc[0, Permutations of shaderName + keywords]  + queueOffset (150, 300 or 450)
+
+Alpha Test 2600 + crc[0, Permutations of shaderName + keywords]  + queueOffset (150, 300 or 450)
+
+Transparent ALWAYS 3000
+```
+
+That algorithm shows some overlapping in some extreme cases.
+- Opaque materials will always be in a render queue < 2600 if we have less than 150 shader variants permutations. A quick test showed that walking around different scenes for an hour accumulated a total of 15 permutations, so assuming `renderQueue < 2600 => Opaque` is safe.
+- Transparent materials will always be in `render queue == 3000` by the algorithm (we dont do further optimization because we need clip space z sorting to render them properly).
```

Comment:
```suggestion
- Transparent materials will always be in `render queue == 3000` by the algorithm (we don't do further optimization because we need clip space z sorting to render them properly).
```